File: backend/src/scripts/createAdmin.js
import 'dotenv/config';
import { connectDB } from '../config/db.js';
import User from '../models/User.js';
import { encryptEmail, hashEmail } from '../utils/crypto.js';

const parseArgs = () => {
  const args = process.argv.slice(2);
  const getArg = (flag) => {
    const index = args.indexOf(flag);
    return index >= 0 ? args[index + 1] : '';
  };

  return {
    email: getArg('--email'),
    username: getArg('--username'),
    password: getArg('--password')
  };
};

const run = async () => {
  const { email, username, password } = parseArgs();

  if (!email || !username || !password) {
    console.error('Usage: npm run create-admin -- --email admin@example.com --username admin --password StrongPass123!');
    process.exit(1);
  }

  await connectDB();

  const normalizedEmail = email.trim().toLowerCase();
  const existing = await User.findOne({
    $or: [{ username: username.trim() }, { emailHash: hashEmail(normalizedEmail) }]
  });

  if (existing) {
    console.error('Admin user already exists with that username or email.');
    process.exit(1);
  }

  const admin = await User.create({
    username: username.trim(),
    emailHash: hashEmail(normalizedEmail),
    emailEncrypted: encryptEmail(normalizedEmail),
    password,
    role: 'admin',
    superiorUntil: new Date('2099-12-31T00:00:00.000Z'),
    sustainabilityScore: 100
  });

  console.log(`Admin created: ${admin.username}`);
  process.exit(0);
};

run().catch((error) => {
  console.error(error);
  process.exit(1);
});

File: backend/app/scripts/create_admin.py
import argparse

from app.main import encrypt_email, get_db, hash_email, hash_password, now_utc


def main() -> None:
    parser = argparse.ArgumentParser(description='Create an admin user.')
    parser.add_argument('--email', required=True)
    parser.add_argument('--username', required=True)
    parser.add_argument('--password', required=True)
    args = parser.parse_args()

    db = get_db()
    email = args.email.strip().lower()
    username = args.username.strip()

    existing = db['users'].find_one({
        '$or': [
            {'username': username},
            {'emailHash': hash_email(email)},
        ],
    })
    if existing:
        print('Admin user already exists for that username or email.')
        return

    timestamp = now_utc()
    db['users'].insert_one({
        'username': username,
        'emailHash': hash_email(email),
        'emailEncrypted': encrypt_email(email),
        'password': hash_password(args.password),
        'role': 'admin',
        'superiorUntil': None,
        'sustainabilityScore': 10,
        'profileImage': '',
        'bio': '',
        'location': 'Peenya Industrial Area',
        'createdAt': timestamp,
        'updatedAt': timestamp,
    })
    print(f'Admin user created: {username}')


if __name__ == '__main__':
    main()
